File: abc333b/abc333b.ml
open Printf
open Scanf

let solve ss tt =
  let to_num x =
    match x with
      'A' -> 0
    | 'B' -> 1
    | 'C' -> 2
    | 'D' -> 3
    | _ -> 4 in
  let dist x y =
    let l = abs (to_num x - to_num y) in
    min l (5 - l) in
  if dist ss.[0] ss.[1] = dist tt.[0] tt.[1] then "Yes" else "No"

let () =
  scanf "%s %s " solve |> printf "%s\n"

File: abc333b/abc333b.hs
import Data.Bool (bool)

main :: IO ()
main = solve <$> getLine <*> getLine >>= putStrLn

solve :: String -> String -> String
solve ss tt = bool "No" "Yes" $ dist ss == dist tt

dist :: String -> Int
dist (x:y:_) = min c (5 - c)
  where
    a = toNum x
    b = toNum y
    c = abs $ a - b

toNum :: Char -> Int
toNum 'A' = 0
toNum 'B' = 1
toNum 'C' = 2
toNum 'D' = 3
toNum _ = 4
